File: 0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode *getIntersectionNode(struct ListNode *headA, struct ListNode *headB) {
    if ((headA == NULL) || (headB == NULL)) {
        return NULL;
    }
    struct ListNode *pointerA = headA, *pointerB = headB;

    while (pointerA != pointerB) {
        pointerA = (pointerA != NULL) ? pointerA->next : headB;
        pointerB = (pointerB != NULL) ? pointerB->next : headA;
    }

    return pointerA;
}

File: Explore Card Track/3. Linked List/0160-intersection-of-two-linked-lists/0160-intersection-of-two-linked-lists.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        # If either head is None, there's no intersection
        if not headA or not headB:
            return None

        # Initialize two pointers for the two lists
        pointerA, pointerB = headA, headB

        # Traverse both lists
        while pointerA != pointerB:
            # Move to the next node in the list, or to the head of the other list if reaching the end
            pointerA = pointerA.next if pointerA else headB
            pointerB = pointerB.next if pointerB else headA

        # Either they met at the intersection node, or both are None (no intersection)
        return pointerA
